### PR TITLE
Improve handling of imperfect profile_json when loading instances from Solr.

### LINKED ISSUE #856: Load_instance_from_solr has issues if model has properties that aren't saved in profile_json
This issue comes up if you add new properties to your model and then try to load existing objects from Solr that don't have the new properties set in profile_json. If the new property is singular (i.e. has ```multiple: false```) then trying to read it will raise an error. If it's a multi-value property then reading it returns nil instead of an empty list. This only happens when you use load_instance_from_solr. If you load from Fedora then it behaves as expected, singular properties return nil and multi-value properties return an empty list.

I understand that arguably this might not really be a bug since you can reasonably be expected to perform some kind of a migration when changing the model. However, it would certainly be handy during development if new properties could be just simply added. And in any case, the behaviour should probably be consistent between loading from Fedora or Solr.

Below is a test case to demonstrate this. Can be placed in loadable_from_json_spec.rb.

```ruby
  context "unset fields in profile" do
    before do
      class UnsetTest < ActiveFedora::Base
      end
      @unset_test_id = UnsetTest.create.id
      Object.send(:remove_const, :UnsetTest)
      class UnsetTest < ActiveFedora::Base
        property :singular_field, multiple: false, predicate: ::RDF::URI('http://example.org/singular')
        property :multiple_field, multiple: true, predicate: ::RDF::URI('http://example.org/multiple')
      end
    end
    after do
      Object.send(:remove_const, :UnsetTest)
    end

    context "when loading from Solr" do
      subject { UnsetTest.load_instance_from_solr(@unset_test_id) }

      it "singular field should be nil" do
        expect(subject.singular_field).to be_nil
      end

      it "multiple field should be []" do
        expect(subject.multiple_field).to eql( [] )
      end
    end

    context "when loading from Fedora" do
      subject { UnsetTest.find(@unset_test_id) }

      it "singular field should be nil" do
        expect(subject.singular_field).to be_nil
      end

      it "multiple field should be []" do
        expect(subject.multiple_field).to eql( [] )
      end
    end

  end
```

And I also made a simple fix that'll pass the test but I'm not entirely confident that this is the right way to fix this or if I'm overlooking something else. I don't know the inner workings of Active Fedora that well. But if these look ok, I can certainly just make this a pull request.

```ruby
------------------- lib/active_fedora/loadable_from_json.rb -------------------
index d3e47b4..a38edbf 100644
@@ -117,7 +117,19 @@ module ActiveFedora
         attached_files[key] = SolrBackedMetadataFile.new
       end
       @resource = SolrBackedResource.new(self.class)
-      self.attributes = attrs.slice(*self.class.attribute_names)
+      new_attributes = {}
+      self.class.attribute_names.each do |a|
+        if !attrs.key? a
+          if self.class.reflect_on_property(a).multiple?
+            new_attributes[a] = []
+          else
+            new_attributes[a] = nil
+          end
+        else
+          new_attributes[a] = attrs[a]
+        end
+      end
+      self.attributes = new_attributes
       # TODO Should we clear the change tracking, or make this object Read-only?
 
       run_callbacks :find
```

### PR DESCRIPTION
**Pull request by ojlyytinen**
Fixes #856

I reorganised the code a bit compared to what I posted in the issue discussion. spec/integration/solr_instance_loader_spec.rb already had similar tests so I moved my tests there instead.

### CODE REVIEW

File: lib/active_fedora/loadable_from_json.rb [hunk 1]
```diff
@@ -105,6 +105,30 @@ def reflection(predicate)
       end
     end
 
+    # Adapt attributes read from Solr to possible minor changes in data model
+    # since the attributes were saved.
+    # @param attrs [Hash] attributes read from Solr
+    # @return [Hash] the adapted attributes
+    def adapt_attributes(attrs)
+      new_attributes = {}
+      self.class.attribute_names.each do |a|
```
Comment (jcoyne) [inline lib/active_fedora/loadable_from_json.rb]: Can you set this `a` variable to something more descriptive? How about `attribute_name`? Also consider using `each_with_object({}) do |attribute_name, new_attributes|` here. Then you can remove the previous line and the last line of the method.

File: lib/active_fedora/loadable_from_json.rb [hunk 2]
```diff
@@ -105,6 +105,30 @@ def reflection(predicate)
...
+        reflection = self.class.reflect_on_property a
+        new_attributes[a] = if !reflection
```
Comment (mjgiarlo) [inline lib/active_fedora/loadable_from_json.rb]: Are lines 116-128 all for setting `new_attributes[a]`? I'd consider either changing the indent to make that obvious or pulling this into a private method of its own for cleaner/leaner separation.

File: lib/active_fedora/loadable_from_json.rb [hunk 3]
```diff
@@ -105,6 +105,30 @@ def reflection(predicate)
...
+          attrs[a] # if this isn't a property, copy value verbatim
+        else
+          multiple=reflection.multiple?
```
Comment (mjgiarlo) [inline lib/active_fedora/loadable_from_json.rb]: Add spaces around the `=`.

File: lib/active_fedora/loadable_from_json.rb (hunk 1 above)
Comment (mjgiarlo) [inline lib/active_fedora/loadable_from_json.rb]: :+1:

File: lib/active_fedora/loadable_from_json.rb [hunk 4]
```diff
@@ -105,6 +105,30 @@ def reflection(predicate)
...
     end
 
+    # Adapt attributes read from Solr to possible minor changes in data model
+    # since the attributes were saved.
+    # @param attrs [Hash] attributes read from Solr
+    # @return [Hash] the adapted attributes
+    def adapt_attributes(attrs)
+      new_attributes = {}
+      self.class.attribute_names.each do |a|
+        reflection = self.class.reflect_on_property a
```
Comment (mjgiarlo) [inline lib/active_fedora/loadable_from_json.rb]: Could you use parentheses to wrap args in method calls?

File: lib/active_fedora/loadable_from_json.rb [hunk 5]
```diff
@@ -105,6 +105,30 @@ def reflection(predicate)
...
+        new_attributes[a] = if !reflection
+          attrs[a] # if this isn't a property, copy value verbatim
+        else
+          multiple=reflection.multiple?
+          if !attrs.key? a
+            # value is missing in attrs, add [] or nil
+            multiple ? [] : nil
+          else
+            # convert an existing scalar to an array if needed
+            (multiple && !attrs[a].is_a?(Array)) ? Array(attrs[a]) : attrs[a]
```
Comment (mjgiarlo) [inline lib/active_fedora/loadable_from_json.rb]: Can this line be simplified to:

```ruby
multiple ? Array.wrap(attrs[a]) : attrs[a]
```

**Comment by ojlyytinen** [comment]
I made the suggested code formatting changes and moved the big if structure into its own private method. And adapt_attributes doesn't really need to be public either so made that private too. And finally rebased.

**Comment by mjgiarlo** [comment]
:+1:

**Comment by mjgiarlo** [comment]
This would benefit from a couple more eyes, too.  @jcoyne @terrellt others?

**Pull request merged**